File: 5-Problem-A-Day/160. Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode next; ListNode(int
 * x) { val = x; next = null; } }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode pA = headA, pB = headB;
        boolean firstA = true, firstB = true;
        while (pA != pB) {
            if (pA == null) {
                if (firstA) {
                    pA = headB;
                    firstA = false;
                } else
                    return null;
            } else
                pA = pA.next;
            if (pB == null) {
                if (firstB) {
                    pB = headA;
                    firstB = false;
                } else
                    return null;
            } else
                pB = pB.next;
        }
        return pB;
    }
}

File: 2023/160. Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *pA = headA, *pB = headB;
        int lenA = 0, lenB = 0;
        while (pA != NULL) {
            lenA++;
            pA = pA->next;
        }
        while (pB != NULL) {
            lenB++;
            pB = pB->next;
        }

        pA = headA, pB = headB;
        while (lenA > lenB) {
            pA = pA->next;
            lenA--;
        }
        while (lenB > lenA) {
            pB = pB->next;
            lenB--;
        }

        while (pA != pB) {
            pA = pA->next;
            pB = pB->next;
        }

        return pA;
    }
};